pkg/controller/sync/sync_controller.go

```diff
@@ -120,13 +174,93 @@ func (r *ReconcileSync) Reconcile(request reconcile.Request) (reconcile.Result,
 		if _, err := r.opa.RemoveData(context.Background(), instance); err != nil {
 			return reconcile.Result{}, err
 		}
+		r.metricsCache.DeleteObject(syncKey)
+		reportMetrics = true
 		return reconcile.Result{}, nil
 	}
 
 	r.log.V(logging.DebugLevel).Info("data will be added", "data", instance)
 	if _, err := r.opa.AddData(context.Background(), instance); err != nil {
+		r.metricsCache.AddObject(syncKey, Tags{
+			Kind:   instance.GetKind(),
+			Status: metrics.ErrorStatus,
+		})
+		reportMetrics = true
+
 		return reconcile.Result{}, err
 	}
 
+	r.metricsCache.AddObject(syncKey, Tags{
+		Kind:   instance.GetKind(),
+		Status: metrics.ActiveStatus,
+	})
+
+	r.metricsCache.addKind(instance.GetKind())
+
+	reportMetrics = true
+
 	return reconcile.Result{}, nil
 }
+
+func NewMetricsCache() *MetricsCache {
+	return &MetricsCache{
+		Cache:      make(map[string]Tags),
+		KnownKinds: make(map[string]bool),
+	}
+}
+
+// need to know encountered kinds to reset metrics for that kind
+// this is a known memory leak
+func (c *MetricsCache) addKind(key string) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	c.KnownKinds[key] = true
+}
+
+func (c *MetricsCache) ResetCache() {
+	c.Cache = make(map[string]Tags)
```

Comment:
This needs to be guarded by a mutex for thread safety.

Reply:
added